avocado/formatters.py

```diff
@@ -95,139 +105,198 @@ def to_FORMAT(value, field=field, concept=self.concept,
+        if output is skip:
+            return skip
 
-            field_formats = list(preferred_formats)
+        # Backwards compat for dicts.
+        if isinstance(output, dict):
+            warn('Formatter methods must should return a tuple '
+                 'of values', DeprecationWarning)
 
-            # Add field type if defined
-            if field:
-                if field.type:
-                    field_formats.append(field.type)
+            return tuple(output.values())
 
-                field_formats.append(field.simple_type)
+        if not isinstance(value, (list, tuple)):
+            return (output,)
 
-            # Fallback to simple type (e.g. number) and finally 'raw'
-            field_formats.append('raw')
+        return output
 
-            for f in field_formats:
-                method = getattr(self, 'to_{0}'.format(f), None)
+    def _process_multiple(self, method, value, kwargs):
+        try:
+            output = method(value,
+                            fields=self.fields,
+                            kwargs=kwargs)
+        except Exception:
+            # Logging can be expensive in loops. Only run in debug mode
+            if settings.DEBUG:
+                if self.concept and self.concept not in self._errors:
+                    self._errors.add(self.concept)
+                    log.exception('Multi-value formatter error')
 
-                if not method:
-                    continue
+            return skip
 
-                try:
-                    value = method(value,
-                                   field=field,
-                                   concept=self.concept,
-                                   process_multiple=False,
-                                   **context)
+        if output is skip:
+            return skip
 
-                    # Add/update output and break loop
-                    if isinstance(value, dict):
-                        output.update(value)
-                    else:
-                        output[key] = value
+        # Backwards compat for dicts.
+        if isinstance(output, dict):
+            warn('Formatter methods must should return a tuple '
+                 'of values', DeprecationWarning)
 
-                    break
-                except Exception:
-                    if field and field not in self._errors:
-                        self._errors.add(field)
-                        log.exception('Single-value formatter error')
+            return tuple(output.values())
 
-        return output
+        if not isinstance(output, (list, tuple)):
+            warn('Formatter output should be a tuple', RuntimeWarning)
 
-    def __contains__(self, choice):
-        return hasattr(self, 'to_{0}'.format(choice))
+            return (output,)
 
-    def __unicode__(self):
-        return u'{0}'.format(self.name or self.__class__.__name__)
+        return output
+
+    def get_fields(self, exporter=None):
```

Comment:
@murphyke This is one incarnation of what we discussed. The idea being that this method should return a sequence of fields representative of the data being produced by this formatter. The reason this is a list of dicts rather than simple strings is to support exporters that want or require more information about the fields, such as Excel (with the data dictionary sheet), SAS and R. It is not sufficient to have the exporter inspect the concept and fields for that content since formatters are allowed to return dynamic content (#99 and #255). The tradeoff is that there is a bit more code to write now for each formatter that returns custom values.

Right now the `exporter` itself is passed in, but I am going to change this to just be the exporter type (as a string) to simplify things. I am going to add a hook to call `get_EXPORTER_fields` if defined, otherwise fallback to the current code. So formatter authors can write smaller methods, isolated to the exporter type.

Thoughts? @tjrivera